bug 1259776 - One click ban hammer: Phase II

dchukhin opened the pull request:
This pull request adds the features described for Phase II of the one-click-ban feature: "Have checkbox to submit revisions to Akismet as spam."

One question that still needs to be answered is: how should we handle invalid data for the `RevisionAkismetSubmissionSpamForm` that we use in [`kuma/users/views.py`](https://github.com/caktus/kuma/blob/7f31d9edaa9dba644939fcece38e49a366c8b3c7/kuma/users/views.py#L155)?
@jwhitlock

jwhitlock commented:
Could you ``rebase -i`` on master and force push? Too much history for me, and it will pick up the ``flake8 `` fix.

Normally, a form error results in the form being reloaded and trying again. The Akismet submission may be a little strange.  Maybe the right thing is a [Django message](https://docs.djangoproject.com/en/1.8/ref/contrib/messages/) saying there was a problem, and link to the  "submit as spam" page in the message. Maybe.

It will be Monday before I can look in detail, have a good weekend.

jwhitlock commented:
Thank you, so much saner.

jwhitlock commented on kuma/users/views.py:
```diff
@@ -23,6 +23,9 @@
 from taggit.utils import parse_tags
 
 from kuma.core.decorators import login_required
+from kuma.core.utils import paginate
```
This is unused, and is causing a real ``flake8`` warning.

jwhitlock commented on kuma/users/views.py:
```diff
@@ -138,6 +141,22 @@ def ban_user_and_cleanup_summary(request, user_id):
+        # TODO: What to do of data is not valid? Currently, fail silently
+        if data.is_valid():
+            data.save()
+
```
We should call this out as requiring further action on the summary page, which is needed for later phases anyway.  After our discussion during the dev meeting, I agree that this work should be done in this PR.

jwhitlock commented on kuma/users/tests/test_views.py:
```diff
@@ -275,9 +282,101 @@ def test_post_banned_user(self):
+
+        # Case 1.) User has no revisions
+        data = {'revision-id': []}
+
```
These three cases should be split into three tests, which probably means creating some helper methods to reduce test duplication.

jwhitlock commented:
I've done an initial review of the backend code, which looks good in general, but haven't done practical tests in my VM, which I'm hoping @stephaniehobson will be able to do.  Please fix the issues and let us know when it's ready for the next round.

stephaniehobson commented:
Some of this feedback is about code that has already been merged, since I'm doing a functionality review not a code review I'm not sure what is in this PR and what is not but it should all be updated. Also, my apologies for the awful direction on the HTML I provided before I left, I think I must have already been on vacation when I wrote it:

Form page:
- [ ] Remove the history column from the table listing revision history (Unless Kadir or John told you to add it? It's not in the mockups, I'm not sure where it came from).
- [ ] Add an empty `<td></td>` as the first child of `<tr class="dashboard-row">`.
- [ ] Add an empty `<td></td>` as the first child of the row containing the diff.
- [ ] Reduce the colspan on the table cell containing the diff to 4.
- [ ] Remove the colspan declaration from `<tr colspan="5">`
- [ ] Check that the 5 steps above result in the table headings lining up over top the appropriate table cells.
- [ ] Change "Because shobsontest has already been banned, you may review and revert their revisions below." to  "shobsontest has already been banned. You may review and revert their revisions below. "
- [ ] Text next to the spam check box should only be "Spam". Do not include information about previous revision in this phase.
- [ ] Button should be labeled "Ban User for Spam & Submit Spam" or "Submit Spam". 
- [ ] Change `class="dashboard-table"` to `class="dashboard-table dashboard-table-spam"` 
- [ ] If a revision has already been submitted to Akismet as spam do not show the checkbox and instead show the text "Already reported as spam."
- [ ] Remove the wrapper div `<div class="wrap">`
- [ ] Remove the wrapper section `<section id="content-main" class="full">`
- [ ] Remove the `id="docs-activity" class="user-section"` from the section wrapping the revision history.

Summary page:
- [ ] Either remove the entire "The following revisions were reverted: " section or do not list any revisions in that section and say "No revisions were reverted".
- [ ] Add a section titled "The following revisions were reported as spam." Only include pages in it that were successfully reported as spam to Akismet.
- [ ] Only include delete link next to pages the spammer created and that have no additional revisions.
- [ ] Only list pages a single time on the summary page even if the spammer made multiple edits or created and edited them.
- [ ] On the section wrapped around the content on the summary page remove the `id="content"` and add `class="text-content readable-line-length"`.
- [ ] Add opening and closing `<ul>` tags around the lists.

dchukhin commented:
Ready for rereview @jwhitlock @stephaniehobson 
As a note, I removed the history column, though it is mentioned in the [Google doc] (https://docs.google.com/document/d/1xewZXZlz-VpUFXff3GQ2QElxgqxFxldN751boGkIBPI/edit#heading=h.8tzwaatgdh5x). However, it is not in the wireframes. Let me know if I need to readd it, or keep it out of the code.

stephaniehobson commented:
Sorry for the delay, this exploded for me when I tried it yesterday and I wrote up a comment report on that.... and then it worked this morning. 

- [ ] Remove text "(Note: No revisions were actually reverted. This feature will be added in Phase 3.) " and say "No revisions were reverted."
- [ ] User should not be able to ban their self, in place of the submit button show a paragraph reading "You cannot ban yourself for spam."
- [ ] Put a `<label>` around the check box and the label `Spam`. 
```
<label>
    <input name="revision-id" value="998884" checked="" type="checkbox">
    <strong>Spam</strong>
</label>
```
- [ ] I'm still seeing pages listed multiple times in "The following revisions were not reverted"

dchukhin commented:
@stephaniehobson can you remember what you did to make the pages show up multiple times in the "The following revisions were not reverted" section of the summary page? I've tried a bunch of different things and can't seem to make that happen.

stephaniehobson commented:
This is what I did to get the page listed twice. I'm not sure what the root cause is:
1) find an existing page to edit
2) edit page with spam account
3) next day, edit page again with spam account

Now that I am looking more closely I also see that the translated page I added spam to is not listed as needing follow up. Maybe what is really happening is that the translation is being listed as the English page?

Here's a screenshot of the actions I took with the test account:
![screen shot 2016-07-05 at 13 24 14](https://cloud.githubusercontent.com/assets/854701/16598669/e37dca64-42b3-11e6-85e3-dea8028d90c2.png)

And how it's being displayed on the follow up page:
![screen shot 2016-07-05 at 13 25 27](https://cloud.githubusercontent.com/assets/854701/16598685/f49519b0-42b3-11e6-9f3b-77afbd4290ae.png)

dchukhin commented:
@stephaniehobson thanks for the detail. Yes, I think that is what is happening, since it is [showing the document title](https://github.com/caktus/kuma/blob/one-click-ban-hammer-phase-2/kuma/users/jinja2/users/ban_user_and_cleanup_summary.html#L76). Is that the expected behavior, or should that be changed to the revision title?

stephaniehobson commented:
The document title is what it should be displaying. However, the links when I hover over the document titles are both to the en-US version.

dchukhin commented:
Hmm. Hovering over the document title should show the [link to the revision] (https://github.com/caktus/kuma/blob/one-click-ban-hammer-phase-2/kuma/users/jinja2/users/ban_user_and_cleanup_summary.html#L75), though the "VIEW PAGE" link should go to the original document.
Where should each revision be linking to?

stephaniehobson commented:
I don't remember where they go and I don't have the VM running but the links for the title, "view page" and "history" each contained "en-US".

stephaniehobson commented:
Here's the HTML I'm getting:
```
<ul>
    <li>
        <a href="/en-US/docs/User:stephaniehobson:test4$revision/998884">
            User:stephaniehobson:test4
        </a>
        <a href="/en-US/docs/User:stephaniehobson:test4" class="button">
            View Page <i aria-hidden="true" class="icon-circle-arrow-right"></i>
        </a>
        <a href="/en-US/docs/User:stephaniehobson:test4$history" class="button">
            History <i aria-hidden="true" class="icon-book"></i>
        </a>
    </li>
    <li>
        <a href="/en-US/docs/User:stephaniehobson:test$revision/998883">
            User:stephaniehobson:test</a>
        <a href="/en-US/docs/User:stephaniehobson:test" class="button">
            View Page <i aria-hidden="true" class="icon-circle-arrow-right"></i>
         </a>
        <a href="/en-US/docs/User:stephaniehobson:test$history" class="button">
                  History <i aria-hidden="true" class="icon-book"></i>
        </a>
    </li>
    <li>
            <a href="/en-US/docs/User:stephaniehobson:test$revision/998882">
                 User:stephaniehobson:test
            </a>
            <a href="/en-US/docs/User:stephaniehobson:test" class="button">
                  View Page
                  <i aria-hidden="true" class="icon-circle-arrow-right"></i>
             </a>
            <a href="/en-US/docs/User:stephaniehobson:test$history" class="button">
                  History
                  <i aria-hidden="true" class="icon-book"></i>
             </a>
         </li>
        <li>
            <a href="/en-US/docs/User:stephaniehobsontest:test$revision/998880">
                  User:stephaniehobsontest:test
            </a>
            <a href="/en-US/docs/User:stephaniehobsontest:test" class="button">
                  View Page
                  <i aria-hidden="true" class="icon-circle-arrow-right"></i>
            </a>
            <a href="/en-US/docs/User:stephaniehobsontest:test$history" class="button">
                  History
                  <i aria-hidden="true" class="icon-book"></i>
            </a>
            <a href="/en-US/docs/User:stephaniehobsontest:test$delete" class="button">
                    Delete
                    <i aria-hidden="true" class="icon-trash"></i>
            </a>
        </li>
</ul>
```

Revision #998882 is the translation. The View Page link for it should be to /fr/docs/User:stephaniehobson:test and the History link should be to /fr/docs/User:stephaniehobson:test$history. Clicking the page title can continue linking to the revision.

jwhitlock commented on kuma/users/jinja2/users/ban_user_and_cleanup_summary.html:
```diff
@@ -24,28 +24,39 @@
 
-        {% for revision in revisions_reverted %}
-          <li>
-            <a href="{{ url('wiki.revision', revision.document.slug, revision.id) }}">{{ revision.title }}</a>
```
Instead of ``{{ url('wiki.revision', ...)``, you should use ``{{ revision.get_absolute_url() }}``, which will take care of locales as well.

jwhitlock commented on kuma/users/tests/__init__.py:
```diff
@@ -67,12 +67,18 @@ def setUp(self):
+        create_new_documents = False
+        if not document:
+            create_new_documents = True
         revisions_created = []
```
``create_new_documents = not document`` fits on one line

jwhitlock commented on kuma/users/jinja2/users/ban_user_and_cleanup.html:
```diff
@@ -26,52 +26,43 @@
+        {% elif not user_banned and not revisions_not_spam %}
+          {{ _('Ban User for Spam') }}
+        {% else %}
+          {{ _('Ban User for Spam & Submit Spam') }}
```
I'm uncomfortable with the naked ampersand, but it works in all the browsers

jwhitlock commented on kuma/users/tests/test_templates.py:
```diff
@@ -484,15 +487,48 @@ def test_not_banned_user_one_click_page_template_ban_button(self):
+        # For self.testuser (not banned, no revisions needing to be reverted)
+        # the button on the form should read "Ban User for Spam". There should
+        # be no link to ban for other reasons
+        # Create some revisions made by self.testuser2 and add a Spam submission for each
```
This comment appears to go with the three ``eq_`` tests above.

jwhitlock commented on kuma/users/tests/test_templates.py:
```diff
@@ -505,27 +541,24 @@ def test_not_banned_user_one_click_page_template_ban_button(self):
         eq_(ban_button.text(), "Ban User for Spam")
         eq_(len(ban_other_reasons), 0)
 
```
While this comment goes with the next section, and is misleading. The revisions are created by ``self.testuser``, submitted as spam by ``self.testuser2``, and then ``self.testuser2`` is loaded for banning. Milliseconds from a spam moderator to getting banned!

This repeated test section seems redundant - it seems to be testing the exact same thing as the previous request / response tests, except also creating some irrelevant revisions for another user. Unless I'm missing something (code coverage?) it seems that the 4-line comment and following code block can be dropped.

dchukhin replied:
Thanks for thinking this through and my apologies for the confusion.
I meant to create some revisions by `self.testuser2`, then to mark them as spam, and then to go to the one-click banning page for `self.testuser2` and to check the text of the ban button, which should say `"Ban User for Spam"`. I've updated the test. Let me know if you still think it needs work.

jwhitlock commented on kuma/users/tests/test_templates.py:
```diff
@@ -505,27 +541,24 @@ def test_not_banned_user_one_click_page_template_ban_button(self):
-        UserBan.objects.create(user=self.testuser2, by=self.admin,
-                               reason='Banned by unit test.',
-                               is_active=True)
 
```
😍 splitting this test

jwhitlock commented on kuma/users/tests/test_templates.py:
```diff
@@ -539,10 +572,25 @@ def test_banned_user_one_click_page_template_ban_button(self):
+        UserBan.objects.create(user=self.testuser, by=self.admin,
+                               reason='Banned by unit test.',
+                               is_active=True)
+        UserBan.objects.create(user=self.testuser2, by=self.admin,
```
``self.testuser`` appears to be absent from the rest of the test, and the ``UserBan`` could be dropped.

dchukhin replied:
good 👀

jwhitlock commented on kuma/users/tests/test_templates.py:
```diff
@@ -560,66 +608,716 @@ def test_banned_user_one_click_page_template_ban_button(self):
+            document=self.document,
+            creator=self.testuser2)
+        for revision in created_revisions:
+            revision.akismet_submissions.add(RevisionAkismetSubmission(sender=self.testuser2, type="spam"))
```
``testuser2`` submits own revisions as spam. Brutal.

jwhitlock commented on kuma/users/tests/test_templates.py:
```diff
@@ -560,66 +608,716 @@ def test_banned_user_one_click_page_template_ban_button(self):
+
 
 @pytest.mark.bans
 class BanUserAndCleanupSummaryTestCase(SampleRevisionsMixin, UserTestCase):
```
I'm assuming all the ``TODO:`` in this section is collaboration on Phase 3 tests.  Normally, I'd want the speculative future code removed, so the PR only covers the changes, but I think this is a special case, so keep them to be uncommented in the next PR.

jwhitlock commented on kuma/users/tests/test_views.py:
```diff
@@ -275,9 +289,112 @@ def test_post_banned_user(self):
+            ok_(submission.revision in revisions_created)
+        # No revisions submitted for self.testuser2, since self.testuser2 had no revisions
+        testuser2_submissions = RevisionAkismetSubmission.objects.filter(revision__creator=self.testuser2.id)
+        eq_(testuser2_submissions.count(), 0)
```
This test for ``testuser2`` submissions seems unnecessary

jwhitlock commented on kuma/users/tests/test_views.py:
```diff
@@ -275,9 +289,112 @@ def test_post_banned_user(self):
+        # No revisions submitted for self.testuser2, since self.testuser2 had no revisions
+        testuser2_submissions = RevisionAkismetSubmission.objects.filter(
+            revision__creator=self.testuser2.id)
+        eq_(testuser2_submissions.count(), 0)
```
Again, unnecessary tests

jwhitlock commented on kuma/users/views.py:
```diff
@@ -138,21 +142,115 @@ def ban_user_and_cleanup_summary(request, user_id):
+        ).annotate(
+            latest_rev=Max('revisions')
+        ).values_list('latest_rev', flat=True)
+    )
```
Nice use of ``Max``. Fancy. 🎆

jwhitlock commented on kuma/users/views.py:
```diff
@@ -138,21 +142,115 @@ def ban_user_and_cleanup_summary(request, user_id):
+            ).annotate(
+                latest_rev=Max('revisions')
+            ).values_list('latest_rev', flat=True)
+        )
```
I like this style more than the 110-column monstrosity of ``revisions_from_last_three_days``.  I've also seen this style, borrowed from the JS crowd, which seems to work well for chained queryset operations:

```
    distinct_submitted_to_akismet_rev_ids = list(
        Document.objects
        .filter(revisions__in=submitted_to_akismet)
        .annotate(latest_rev=Max('revisions'))
        .values_list('latest_rev', flat=True))
```

dchukhin replied:
looks readable to me

jwhitlock commented:
I didn't find any issues in my brief practical tests. It looks good, and will be useful.  I'll need to do a little more practical testing in the morning, but it looks like it will work from the code.

My remaining review tasks for the morning:
* [ ] Multiple edits to a single page, page appears once on summary
* [ ] Spam edit on non-English page, link is to proper locale

There's some cleanup to make the tests more focused:
* [ ] ``test_not_banned_user_no_revisions_or_already_spam_ban_button`` in ``kuma/users/tests/test_templates.py`` - delete code that tests again with ``testuser2``
* [ ] ``test_banned_user_no_revisions_or_already_spam_ban_button`` in ``test_templates`` - don't ban ``testuser``
* [ ] ``test_post_submits_revisions_to_akismet_as_spam`` in ``kuma/users/tests/test_views.py`` - drop tests that ``testuser2``'s nonexistent revisions have not been submitted.
* [ ] ``test_post_submits_no_revisions_to_akismet_when_no_user_revisions`` in ``test_views.py`` - again, drop ``testuser2`` tests

And then there's some completely optional style stuff:
* [ ] ``create_revisions`` in ``kuma/users/tests/__init__.py`` - one-line assignment of ``create_new_documents``
* [ ] JS-style chained operations for querysets (watch out for ``flake8``)

dchukhin commented:
@jwhitlock I've implemented the changes noted above, except for this one:
 - `test_not_banned_user_no_revisions_or_already_spam_ban_button` in `kuma/users/tests/test_templates.py` - delete code that tests again with testuser2
Instead I changed the test to first look at the ban button for `self.testuser` (no revisions), then for `self.testuser2` (revisions already marked spam), which I considered 2 different situations. Let me know if that is still redundant.

jwhitlock commented:
Yes, I agree with the re-written ``test_not_banned_user_no_revisions_or_already_spam_ban_button``. Thanks!
